cql3/statements/alter_session_statement.cc

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Copyright (C) 2015 ScyllaDB
+ *
+ * Modified by ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "cql3/statements/alter_session_statement.hh"
+
+#include "transport/messages/result_message.hh"
+
+namespace cql3 {
+
+namespace statements {
+
+alter_session_statement::alter_session_statement(::shared_ptr<alter_session_prop_defs> props)
+        : cql_statement_no_metadata(&timeout_config::other_timeout)
+        , props(props)
+{
+}
+
+uint32_t alter_session_statement::get_bound_terms() const {
+    return 0;
+}
+
+bool alter_session_statement::depends_on_keyspace(const sstring& ks_name) const
+{
+    return false;
+}
+
+bool alter_session_statement::depends_on_column_family(const sstring& cf_name) const
+{
+    return false;
+}
+
+future<> alter_session_statement::check_access(service::storage_proxy& proxy, const service::client_state& state) const
+{
+    state.validate_login();
+    return make_ready_future<>();
+}
+
+void alter_session_statement::validate(service::storage_proxy&, const service::client_state& state) const
+{
+    props->validate();
+}
+
+future<::shared_ptr<cql_transport::messages::result_message>>
+alter_session_statement::execute(service::storage_proxy& proxy, service::query_state& state, const query_options& options) const {
+    state.get_client_state().set_session_params(props->get_params());
+    auto result = ::make_shared<cql_transport::messages::result_message::void_message>();
```

Comment:
How is the driver going to find out that session properties were altered? Will it have to parse CQL to find out? 
I though you were going to return a metadata flag indicating that session properties were changed.

How do I retreive session properties using CQL?

Reply:
It's kept in client state which as far as I remember has a CQL table that stores the info for all client states.

Reply:
That's how I imagined it, given that parsing this statement is a bunch of string splits. We could even go one step further and represent it as JSON for free parsing.

As for retrieving the info - I don't have a good idea. How does SQL do that, with a local (virtual) table?

Reply:
Oh, right, we have `system.clients` now, which is a very good fit.

Reply:
I think the drivers should not have to parse CQL to track session changes.

Reply:
I generally agree, but modifying the binary protocol is also a mess, since it requires the driver and the server to agree on who understands what. In case of the driver side, it could actually be made even simpler (and require no parsing), if the session altering is never sent in a raw form, but rather built by the driver itself, e.g. from a native map implementation (something like that in pseudopython: `session.execute(AlterSessionStatement({'foo': 'bar', 'foo2': 'baz'}))`).
That's how `USE <keyspace>` works in gocql, as far as I understood from my consultation with PiotrD - sending raw `USE` statements is disallowed, and the only way to use a keyspace is to set it up via gocql API.

Reply:
Ah. But what about cqlsh?

Reply:
cqlsh parses some statements:
```python
        if custom_handler:
            parsed = cqlruleset.cql_whole_parse_tokens(tokens, srcstr=srcstr,
                                                       startsymbol='cqlshCommand')
```
and then
```python
    def do_use(self, parsed):
        ksname = parsed.get_binding('ksname')
        success, _ = self.perform_simple_statement(SimpleStatement(parsed.extract_orig()))
        if success:
            if ksname[0] == '"' and ksname[-1] == '"':
                self.current_keyspace = self.cql_unprotect_name(ksname)
            else:
                self.current_keyspace = ksname.lower()
```

Reply:
> I generally agree, but modifying the binary protocol is also a mess, since it requires the driver and the server to agree on who understands what. In case of the driver side, it could actually be made even simpler (and require no parsing), if the session altering is never sent in a raw form, but rather built by the driver itself, e.g. from a native map implementation (something like that in pseudopython: `session.execute(AlterSessionStatement({'foo': 'bar', 'foo2': 'baz'}))`).
> That's how `USE <keyspace>` works in gocql, as far as I understood from my consultation with PiotrD - sending raw `USE` statements is disallowed, and the only way to use a keyspace is to set it up via gocql API.

I agree. Another option, which I think is inferior, is to send ALTER SESSION and have the server respond with a flag that says "this needs to go to all connections". The driver sees the flag and broadcasts the alter.